### PR TITLE
Add server operation registry and router

### PR DESCRIPTION
**Pull request by david-perez**
This commit adds two things:
1. A runtime router implementing `tower`'s [`Service`](https://docs.rs/tower-service/0.3.1/tower_service/trait.Service.html) that adheres to [Smithy's `http` trait specification](https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html#http-trait), that is linear in the number of registered routes.
2. A code-generated "operation registry" that allows service implementers to provide Rust functions and declare them as the handlers for their service's operations.

The framework will receive HTTP requests from the server and route them to the corresponding operation handler.

---

~~The PR is far from being mergeable, but it builds. I also want to gather feedback, especially for `OperationRegistry.kt`.~~ PR is now mergeable.

Another PR implementing `axum`'s `FromRequest` and `IntoResponse` for operations needs to be merged first, ~~on which @crisidev is currently working.~~ which was done on #839.

### CODE REVIEW

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt
Comment (rcoh) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt:43]: you can actually just `Attribute.Custom(...).render()`. Raw is also fine, totally up to you

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt
Comment (rcoh) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt:44]: as in a human readable name of the service? _probably_ not, you probably want something out of `RustSettings`

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt
Comment (rcoh) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt:158]: 1. yes, the main reason to use `RuntimeType` is so that you ensure you add the dependency, but only when it's actually used. As the project gets more complex, you may end up with certain services using differing subsets of dependencies
2. yes, you can call `addDependency` directly on the writer

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/CargoDependency.kt
Comment (rcoh) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/CargoDependency.kt:221]: this should be done on main

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt
Comment (david-perez) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt:44]: This gives me the service name defined in the Smithy model file, which should be appropiate here

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt
Comment (crisidev) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt:84]: Same as above

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt
Comment (crisidev) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt:138]: This can go away now!

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/CargoDependency.kt
Comment (crisidev) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/CargoDependency.kt:216]: Why is this gone?

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/RuntimeTypes.kt
Comment (crisidev) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/RuntimeTypes.kt:187]: Oh, this are much nicer that what I have used in the IntoResponse/FromRequest implementation. We can change them there!

File: rust-runtime/aws-smithy-http-server/Cargo.toml
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/Cargo.toml:4]: Please leave the alias instead of names. It will make things easier in the future as the email list is already active and contains me and you right now.

File: rust-runtime/aws-smithy-http-server/Cargo.toml
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/Cargo.toml:38]: Is this TODO already applicable?

File: rust-runtime/aws-smithy-http-server/src/clone_box_service.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/clone_box_service.rs:1]: Missing SPDX licence header and module description for Rust doc. If this comes from axum with modification, you will need to follow what I did in `error.rs` and add the axum license inline.

Comment (crisidev) [reply rust-runtime/aws-smithy-http-server/src/clone_box_service.rs:1]: I would also add a description of why this module is here and how it is used in the codebase as it is not the most immediate code we have.

File: rust-runtime/aws-smithy-http-server/src/extractors.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/extractors.rs:1]: This can go away since my PR has been merged.

File: rust-runtime/aws-smithy-http-server/Cargo.toml
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/Cargo.toml:30]: From pin-project-lite documentation:
```This is the only reason to use this crate. However, if you already have proc-macro related dependencies in your crate's dependency graph, there is no benefit from using this crate. (Note: There is almost no difference in the amount of code generated between pin-project and pin-project-lite.)```

We already have proc-macro deps coming in with `async-trait` crate. Is there any other reason to use this instead of standard `pin-project` ?

File: rust-runtime/aws-smithy-http-server/Cargo.toml
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/Cargo.toml:34]: Tokio is now at 1.13.0. Any reason to not use "1" so we get the latest stable version?

File: rust-runtime/aws-smithy-http-server/src/handler/mod.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/handler/mod.rs:1]: Missing SPDX licence header and module description for Rust doc.

File: rust-runtime/aws-smithy-http-server/src/lib.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/lib.rs:5]: Please keep the SPDX license identifier header.

File: rust-runtime/aws-smithy-http-server/src/lib.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/lib.rs:7]: This can go away right?

File: rust-runtime/aws-smithy-http-server/src/macros.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/macros.rs:177]: If you add a small doc about what is an opaque future, it will reflect into cargo docs, which is nice.

File: rust-runtime/aws-smithy-http-server/src/main.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/main.rs:1]: This can go away or should be moved into an example if we want to keep it, right? 

Missing SPDX licence header.

File: rust-runtime/aws-smithy-http-server/src/responders.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/responders.rs:1]: This can go away before merging the PR.

File: rust-runtime/aws-smithy-http-server/src/routing/future.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/routing/future.rs:1]: Missing SPDX licence header.

File: rust-runtime/aws-smithy-http-server/src/runtime.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/runtime.rs:1]: This can go away before merging the PR. It is already implemented.

File: rust-runtime/aws-smithy-http-server/src/routing/route.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/routing/route.rs:1]: Missing SPDX licence header and module description for Rust doc. If this comes from axum with modification, you will need to follow what I did in error.rs and add the axum license inline.

**Review by crisidev — CHANGES_REQUESTED** [review]
I still need to go deeper into the rust implementation of your PR, but I have published some comments you can look at and work on already!

Nice work so far.

File: rust-runtime/aws-smithy-http-server/src/model.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/model.rs:1]: This can go away as it is autogenerated now from your previous PR.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt
Comment (david-perez) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt:31]: In the rest of the codebase we're using `symbolProvider.toSymbol(operationShape).name`. I'm changing it to that to be consistent. I wonder  what the preferred way of doing this is.

File: rust-runtime/aws-smithy-http-server/Cargo.toml
Comment (david-perez) [reply rust-runtime/aws-smithy-http-server/Cargo.toml:38]: Yes, it was here to have the example application in `main.rs` of the `aws-smithy-http-server` build.

File: rust-runtime/aws-smithy-http-server/Cargo.toml
Comment (david-perez) [reply rust-runtime/aws-smithy-http-server/Cargo.toml:34]: Isn't the behavior the same? From https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#caret-requirements

> An update is allowed if the new version number does not modify the left-most non-zero digit in the major, minor, patch grouping.

File: rust-runtime/aws-smithy-http-server/src/main.rs
Comment (david-perez) [reply rust-runtime/aws-smithy-http-server/src/main.rs:1]: I've removed it. We can move it to an example, but it's going to require us manually writing extractors, responders, model, operation registry (essentially everything that we code-generate). So perhaps it's best we do examples as unit tests in `codegen-server`.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt
Comment (david-perez) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt:57]: Yeah I don't really know if I like it. I saw it used in the codebase and copied it. But, IMO

```kotlin
rustTemplate("""
    pub struct #{Foo} {
        bar: #{Bar},
    }""", *codegenScope)
```

reads better than:

```kotlin
rustBlockTemplate("pub struct #{Foo}", *codegenScope) {
   rustTemplate("bar: #{Bar},", *codegenScope)
}
```

This is another instance where I think there are too many ways in the codebase to generate the same thing and we need simple rules of thumb to homogenize. Perhaps we can discuss with @rcoh. Ideally we'd have all the methods in `RustWriter.kt` documenting when we should favor one over the other.

Comment (rcoh) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/OperationRegistryGenerator.kt:57]: I generally agree that not using a nested block is more readable—it's only in the places where you need to loop over something that the second pattern is clearer.

**Comment by david-perez** [comment]
I have resolved all the comments that I have addressed; ready for a second review round.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGenerator.kt
Comment (crisidev) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGenerator.kt:39]: Out of curiosity what is wrong with this map?

File: rust-runtime/aws-smithy-http-server/Cargo.toml
Comment (crisidev) [reply rust-runtime/aws-smithy-http-server/Cargo.toml:34]: There is no caret here, it's just "1.0" so for what I understand no, the behaviour is not the same.

Comment (david-perez) [reply rust-runtime/aws-smithy-http-server/Cargo.toml:34]: From https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-cratesio

> The string "0.1.12" is a semver version requirement. Since this string does not have any operators in it, it is interpreted the same way as if we had specified "^0.1.12", which is called a caret requirement.

File: rust-runtime/aws-smithy-http-server/Cargo.toml
Comment (david-perez) [reply rust-runtime/aws-smithy-http-server/Cargo.toml:30]: Nope. Switched to `pin-project`.

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/CargoDependency.kt
Comment (david-perez) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/CargoDependency.kt:216]: We weren't using it.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGenerator.kt
Comment (david-perez) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGenerator.kt:39]: A pet peeve of mine: `.map` should be used to map a value to another value _without side-effects_. As written this returns `Unit` and has lots of side effects. A for loop communicates intent better.

Comment (crisidev) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGenerator.kt]: Got it.

File: rust-runtime/aws-smithy-http-server/Cargo.toml
Comment (crisidev) [reply rust-runtime/aws-smithy-http-server/Cargo.toml:34]: Ah! Ok, forget what I said here.

**Review by crisidev — APPROVED** [review]
LGTM

**Comment by david-perez** [comment]
Resolved some clippy lints; hopefully CI passes now.

**Review by crisidev — APPROVED** [review]

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/CargoDependency.kt
Comment (rcoh) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/CargoDependency.kt:195]: in the long term, you _probably_ just want to code generate builders directly, I suspect. `syn` is a very large dependency that ideally, won't be a dependency of the generated code

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/RuntimeTypes.kt
Comment (rcoh) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/RuntimeTypes.kt:185]: may have mentioned this a while ago, but you may at some point want your own companion object in `codegen-server` that contains server-specific runtime types

**Review by rcoh — APPROVED** [review]

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/CargoDependency.kt
Comment (david-perez) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/CargoDependency.kt:195]: Good point. https://github.com/awslabs/smithy-rs/issues/864

**Pull request merged**
